independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java

```diff
@@ -55,8 +70,33 @@ public void walk(PathVisitor visitor) {
         PathTreeVisit.walk(dir, dir, pathFilter, getMultiReleaseMapping(), visitor);
     }
 
+    private String ensureRelative(String path) {
+        if (path == null || path.isEmpty()) {
+            return path;
+        }
+        if (isWindowsAbsolutePath(path)) {
+            // We are allowing absolute paths on Linux but interpreting the root as the root of the tree.
+            // However, Windows absolute paths, including the disk part, are not expected.
+            throw new IllegalArgumentException(
+                    path + " does not appear to be a path relative to the root of the path tree " + dir);
+        }
+        // this is to disallow reading outside the path tree root
+        if (path.contains("..")) {
+            final Path absolutePath = dir.resolve(path).normalize().toAbsolutePath();
+            if (absolutePath.startsWith(dir)) {
+                return dir.relativize(absolutePath).toString();
+            }
+            return null;
+        }
+        if (path.charAt(0) == '/') {
+            return path.substring(1);
```

Comment:
We should still disallow reading outside the tree.

Reply:
Yes, that's what I meant. Perhaps after stripping the first `/` I could check again if it still starts with `/` and then throw an IAE.  I suppose it wouldn't be a legal resources name.

Reply:
It doesn't appear that a resource name can start with `/`. I just tried
````
		Enumeration<URL> resources = Thread.currentThread().getContextClassLoader().getResources("/META-INF");
````
in a non-Quarkus test and it returns an empty enumeration, while it returns 177 URLs for `"META-INF"`. So it looks like we could simply disallow absolute paths.

Reply:
It could potentially break existing callers/users though.

Reply:
> Perhaps after stripping the first `/` I could check again if it still starts with `/` and then throw an IAE. I suppose it wouldn't be a legal resources name.

I pushed this adjustment. Although I'm still wondering whether we should simply disallow absolute paths completely. Perhaps, simply returning `null`, indicating "it's not visible from within the tree".

Reply:
Yes, I think we should handle absolute paths as not visible. Do you agree @yrodiere ?

Reply:
> It doesn't appear that a resource name can start with `/`.

It depends. It can if you use `java.lang.Class#getResource` to retrieve the resource, because by default the path is interpreted as relative to the class. It cannot start with `/` if you use `java.lang.ClassLoader#getResource`, because then the resource name is absolute by default.

But I think it's safe to assume we'll always use `ClassLoader#getResource`; I would personally consider it's a bug if any extension/library uses `Class#getResource` instead.

> Yes, I think we should handle absolute paths as not visible. Do you agree @yrodiere ?

Works for me. As long as users have an actionable error message, it should be fine.

Reply:
Ah, you meant to return `null`. I'm not sure that's great, because it will leave users wondering why `/META-INF/MANIFEST.MF` tells them the file doesn't exist, while it works when they do `MyClass.getResource("/META-INF/MANIFEST.MF")`, then it works. I know it's not our fault, it's a quirk of Java APIs, but still, it would be nice to avoid that kind of confusion.

Don't you think it would be fine to throw an exception here, because we consider a resource path starting with `/` is invalid? Something like `new ConfigurationError( "Resource paths must not start with '/'" )`. Then users know what's wrong and can solve it very easily.

Reply:
Then I guess we could leave it as it is.

Reply:
My idea was to make consistent with ClassLoader.getResource(name). Tbh, I haven't seen people using Class.getResource(name). It's looking strange to me they don't have the same resource path rules.

Reply:
> Then I guess we could leave it as it is.

Well the current code will let absolute paths slip through and resolve files outside of the tree.

But you could indeed use something very close to your code. Instead of failing on Windows absolute paths, just fail on all absolute paths, and don't try to handle them at all. Simple and easy.

Reply:
> Did you see an adjustment I pushed yesterday night? Could you give an example of an absolute path that would slip through?

Ah sorry, I was looking at [ArchivePathTree.java](https://github.com/quarkusio/quarkus/pull/23692/files#diff-0a3f8c0e980f0271b0467b82cba352b591567c97d92ca6b82b6ade7c7d0aeff8), which you didn't adjust, apparently.

So the current code is this:

```java
        if (path.charAt(0) == '/') {
            if (path.length() == 1) {
                return "";
            }
            if (path.charAt(1) == '/') {
                throw new IllegalArgumentException(
                        path + " does not appear to be a path relative to the root of the path tree " + dir);
            }
            return path.substring(1);
        }
```

This solution would have been secure: it allows `/foo/bar` and interprets it as `foo/bar`. It rejects `//foo/bar` and `C:\\foo\bar` with an exception.

However... as you very rightly pointed out, `ClassLoader#getResource` doesn't accept leading slashes. If we do here, we run the risk of accepting resource paths such as `/foo/bar` in build steps, then failing to retrieve the same resource during static init (when calling `ClassLoader#getResource` with the same resource path).

So... it's probably better to just reject all absolute paths?

> I like the idea of not supporting absolute paths. In other cases though, if a path isn't on the tree I let the caller decide whether to throw an exception or ignore. Which is why I'd probably prefer to do the same for absolute paths.

Returning `null` when the user sets an absolute path doesn't seem great to me, because then the caller cannot differentiate between a path that points to a file that doesn't exist (i.e. `foo/doesnotexist.txt`) and a path that is simply incorrect but would point to an existing file if it was relative (i.e. `/foo/exists.txt`). This leads to scenarios where the user sets `/foo/exists.txt`, and gets an error message telling him the file doesn't exist. But it does exist, it's just that we don't understand the path, which is a very different situation!

So, same conclusion: it's probably better to just reject all absolute paths? With a clear exception message telling users that we don't accept absolute paths and they should just make their path relative to their resources (e.g. the `src/main/resources` directory).

Reply:
Ok @yrodiere, done